File: Frontend/game/game.js
// 전역 변수
let selectedPlayers = 1; // 기본값 1인
let gameStarted = false;
let gameHistory = []; // 게임 결과 기록
let currentGameData = null; // 현재 게임 데이터 저장

// DOM이 로드된 후 실행
document.addEventListener("DOMContentLoaded", function () {
  // 플레이어 수 선택 버튼 이벤트
  const playerBtns = document.querySelectorAll(".player-btn");
  playerBtns.forEach((btn) => {
    btn.addEventListener("click", function () {
      // 이미 게임이 시작된 경우 변경 불가
      if (gameStarted) {
        alert(
          "게임이 이미 시작되었습니다. 초기화 버튼을 눌러 새 게임을 시작하세요."
        );
        return;
      }

      // 기존 선택 제거
      playerBtns.forEach((b) => b.classList.remove("active"));

      // 새 선택 추가
      this.classList.add("active");

      // 선택된 플레이어 수 저장
      selectedPlayers = parseInt(this.getAttribute("data-players"));

      // 플레이어 영역 초기화
      initializePlayerArea();
    });
  });

  // 기본 선택 (1인)
  playerBtns[0].classList.add("active");
  initializePlayerArea();

  // 게임 시작 버튼 이벤트
  document.getElementById("start-btn").addEventListener("click", startGame);

  // 초기화 버튼 이벤트
  document.getElementById("reset-btn").addEventListener("click", resetGame);

  // 결과 기록 버튼 이벤트
  document.getElementById("history-btn").addEventListener("click", showHistory);

  // 모달 닫기 버튼 이벤트
  document.querySelector(".close-btn").addEventListener("click", function () {
    document.getElementById("history-modal").style.display = "none";
  });

  // 모달 외부 클릭 시 닫기
  window.addEventListener("click", function (event) {
    const modal = document.getElementById("history-modal");
    if (event.target === modal) {
      modal.style.display = "none";
    }
  });
  // BGM 컨트롤 초기화
  initializeAudioControls();
});

// 플레이어 영역 초기화
function initializePlayerArea() {
  const playerContainer = document.getElementById("player-container");
  playerContainer.innerHTML = "";

  const username = localStorage.getItem("username") || "플레이어 1";

  for (let i = 1; i <= selectedPlayers; i++) {
    const playerBox = document.createElement("div");
    playerBox.className = "player-box";
    playerBox.id = `player-${i}`;

    // 플레이어 이름 설정
    const playerName = i === 1 ? username : `플레이어 ${i}`;

    playerBox.innerHTML = `
      <h3>${playerName}</h3>
      <div class="cards">
          <div class="card">
              <img src="cards/card_back.png" alt="카드 뒷면" class="card-image">
          </div>
          <div class="card">
              <img src="cards/card_back.png" alt="카드 뒷면" class="card-image">
          </div>
      </div>    
      <div class="pattern-result">-</div>
      <div class="winner-badge">승리!</div>
    `;

    playerContainer.appendChild(playerBox);
  }
}

// 게임 시작
function startGame() {
  if (gameStarted) {
    alert(
      "게임이 이미 시작되었습니다. 초기화 버튼을 눌러 새 게임을 시작하세요."
    );
    return;
  }

  gameStarted = true;
  document.getElementById("start-btn").disabled = true;
  document.getElementById("reset-btn").disabled = false;

  // API 호출하여 게임 시작
  fetch("/api/game/start", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ players: selectedPlayers }),
  })
    .then((response) => response.json())
    .then((data) => {
      // 게임 결과 표시
      displayGameResults(data);

      // 현재 게임 데이터 저장
      currentGameData = data;

      // 게임 기록 저장
      saveGameHistory(data);
    })
    .catch((error) => {
      console.error("게임 시작 오류:", error);
      // 오류 발생 시 프론트엔드에서 임시 데이터 생성 (개발용)
      const mockData = generateMockGameData();
      displayGameResults(mockData);
      currentGameData = mockData;
      saveGameHistory(mockData);
    });
}

// 게임 초기화 (카드 재배분 없음)
function resetGame() {
  // 플레이어 영역 초기화 (카드 숨김)
  const playerBoxes = document.querySelectorAll(".player-box");
  playerBoxes.forEach((box) => {
    // 승자 클래스 제거
    box.classList.remove("winner");

    const cardImages = box.querySelectorAll(".card-image");
    cardImages.forEach((img) => {
      img.src = "cards/card_back.png";
      img.alt = "카드 뒷면";
    });
    box.querySelector(".pattern-result").textContent = "-";
  });

  // 게임 결과 영역 초기화
  const gameResult = document.getElementById("game-result");
  gameResult.style.display = "none";
  gameResult.textContent = "";
  gameResult.className = "game-result";

  // 게임 상태 초기화
  gameStarted = false;
  document.getElementById("start-btn").disabled = false;
  document.getElementById("reset-btn").disabled = true;

  // API 호출하여 게임 상태 초기화 (서버에서는 세션만 초기화)
  fetch("/api/game/reset", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ players: selectedPlayers }),
  }).catch((error) => {
    console.error("게임 초기화 오류:", error);
  });
}

// 결과 기록 표시
function showHistory() {
  const historyContainer = document.getElementById("history-container");
  historyContainer.innerHTML = "";

  // API 호출하여 게임 기록 가져오기
  fetch("/api/game/history")
    .then((response) => response.json())
    .then((data) => {
      renderHistoryItems(data);
    })
    .catch((error) => {
      console.error("게임 기록 조회 오류:", error);
      // 오류 발생 시 로컬 기록 사용
      renderHistoryItems(gameHistory);
    });

  // 모달 표시
  document.getElementById("history-modal").style.display = "block";
}

// 게임 결과 표시
function displayGameResults(data) {
  const username = localStorage.getItem("username") || "플레이어 1"; // ✅ 여기 추가

  data.players.forEach((player, index) => {
    const playerBox = document.getElementById(`player-${index + 1}`);
    if (!playerBox) return;

    // ✅ 플레이어 이름 설정
    const playerName = index === 0 ? username : `플레이어 ${index + 1}`;
    playerBox.querySelector("h3").textContent = playerName; // ✅ 여기에서 이름 적용

    // 승자 표시
    if (player.isWinner) {
      playerBox.classList.add("winner");
    } else {
      playerBox.classList.remove("winner");
    }

    const cardImages = playerBox.querySelectorAll(".card-image");
    const patternResult = playerBox.querySelector(".pattern-result");

    // 카드 이미지 표시
    cardImages[0].src = `cards/card_${player.cards[0].month}_${player.cards[0].type}.png`;
    cardImages[0].alt = `${player.cards[0].month}월 ${
      player.cards[0].type === 1 ? "광" : "열"
    }`;

    cardImages[1].src = `cards/card_${player.cards[1].month}_${player.cards[1].type}.png`;
    cardImages[1].alt = `${player.cards[1].month}월 ${
      player.cards[1].type === 1 ? "광" : "열"
    }`;

    // 족보 결과 표시
    patternResult.textContent = player.pattern;
  });

  // 게임 결과 영역에 승자 표시
  const gameResult = document.getElementById("game-result");
  gameResult.style.display = "block";

  if (data.isDraw > 0) {
    gameResult.textContent = "무승부!";
    gameResult.className = "game-result draw";
  } else if (data.isDraw < 0) {
    gameResult.textContent = "구사패 재경기!";
    gameResult.className = "game-result draw";
  } else if (data.winner >= 0) {
    gameResult.textContent = `플레이어 ${data.winner + 1} 승리!`;
    gameResult.className = "game-result winner";
  } else {
    gameResult.style.display = "none";
  }
}

// 게임 기록 저장
function saveGameHistory(data) {
  // 최대 10개까지만 저장
  if (gameHistory.length >= 10) {
    gameHistory.shift(); // 가장 오래된 기록 제거
  }

  // 타임스탬프 추가
  data.timestamp = new Date().toLocaleString();

  // ✅ 사용자 이름을 첫 번째 플레이어에 주입
  const username = localStorage.getItem("username") || "플레이어 1";
  if (data.players && data.players.length > 0) {
    data.players[0].username = username;
  }

  // 기록 추가
  gameHistory.push(data);
}

// 기록 항목 렌더링
function renderHistoryItems(historyData) {
  const historyContainer = document.getElementById("history-container");

  if (historyData.length === 0) {
    historyContainer.innerHTML = "<p>게임 기록이 없습니다.</p>";
    return;
  }

  // 최근 기록부터 표시 (최대 10개)
  const recentHistory = historyData.slice(-10).reverse();

  recentHistory.forEach((game, gameIndex) => {
    const historyItem = document.createElement("div");
    historyItem.className = "history-item";

    // 게임 헤더 (번호, 시간)
    const historyHeader = document.createElement("div");
    historyHeader.className = "history-header";
    historyHeader.innerHTML = `
            <span>게임 #${recentHistory.length - gameIndex}</span>
            <span>${game.timestamp || "기록 없음"}</span>
        `;

    // 승자 또는 무승부 정보 추가
    let winnerInfo = "";
    if (game.isDraw > 0) {
      winnerInfo = " (무승부)";
    } else if (game.isDraw < 0) {
      winnerInfo = " (구사패 재경기!)";
    } else if (game.winner >= 0) {
      winnerInfo = ` (플레이어 ${game.winner + 1} 승리)`;
    }

    historyHeader.innerHTML = `
            <span>게임 #${recentHistory.length - gameIndex}</span>
            <span>${game.timestamp || "기록 없음"}${winnerInfo}</span>
        `;

    // 플레이어 결과
    const historyPlayers = document.createElement("div");
    historyPlayers.className = "history-players";

    game.players.forEach((player, playerIndex) => {
      const historyPlayer = document.createElement("div");
      historyPlayer.className = player.isWinner
        ? "history-player winner"
        : "history-player";

      // ✅ 이름 우선순위: 저장된 username > 플레이어 n
      const playerName = player.username || `플레이어 ${playerIndex + 1}`;

      historyPlayer.innerHTML = `
        <div class="history-cards">
          <div class="history-card">
            <img src="cards/card_${player.cards[0].month}_${
        player.cards[0].type
      }.png" 
                 alt="${player.cards[0].month}월 ${
        player.cards[0].type === 1 ? "광" : "열"
      }" 
                 class="history-card-image">
          </div>
          <div class="history-card">
            <img src="cards/card_${player.cards[1].month}_${
        player.cards[1].type
      }.png" 
                 alt="${player.cards[1].month}월 ${
        player.cards[1].type === 1 ? "광" : "열"
      }" 
                 class="history-card-image">
          </div>
        </div>
        <div class="pattern-result">${playerName}: ${player.pattern}</div>
        ${player.isWinner ? '<div class="winner-badge">승리!</div>' : ""}
      `;

      historyPlayers.appendChild(historyPlayer);
    });

    historyItem.appendChild(historyHeader);
    historyItem.appendChild(historyPlayers);
    historyContainer.appendChild(historyItem);
  });
}

// BGM 컨트롤 초기화
function initializeAudioControls() {
  // BGM 요소
  bgmPlayer = document.getElementById("bgm");

  // 볼륨 설정
  bgmPlayer.volume = 0.5;

  // 재생 버튼
  document.getElementById("play-btn").addEventListener("click", function () {
    bgmPlayer.play();
    this.classList.add("active");
    document.getElementById("pause-btn").classList.remove("active");
  });

  // 일시정지 버튼
  document.getElementById("pause-btn").addEventListener("click", function () {
    bgmPlayer.pause();
    this.classList.add("active");
    document.getElementById("play-btn").classList.remove("active");
  });

  // 음소거 버튼
  document.getElementById("mute-btn").addEventListener("click", function () {
    if (bgmPlayer.muted) {
      bgmPlayer.muted = false;
      this.textContent = "🔊";
    } else {
      bgmPlayer.muted = true;
      this.textContent = "🔇";
    }
  });

  // 볼륨 슬라이더
  document
    .getElementById("volume-slider")
    .addEventListener("input", function () {
      bgmPlayer.volume = this.value / 100;
    });

  // 페이지 로드 시 자동 재생 (사용자 상호작용 필요)
  document.addEventListener(
    "click",
    function autoPlayHandler() {
      bgmPlayer.play().catch((error) => {
        console.log("자동 재생 실패:", error);
      });
      document.removeEventListener("click", autoPlayHandler);
    },
    { once: true }
  );
}

// 개발용 임시 데이터 생성 함수 (백엔드 연동 전 테스트용)
function generateMockGameData() {
  const cardTypes = [1, 2]; // 1: 광, 2: 열
  const patternNames = [
    "38광땡",
    "장땡",
    "9땡",
    "알리",
    "독사",
    "구삥",
    "장삥",
    "장사",
    "7끗",
    "4끗",
  ];

  const players = [];

  for (let i = 0; i < selectedPlayers; i++) {
    // 랜덤 카드 생성
    const card1 = {
      month: Math.floor(Math.random() * 10) + 1,
      type: cardTypes[Math.floor(Math.random() * cardTypes.length)],
    };

    const card2 = {
      month: Math.floor(Math.random() * 10) + 1,
      type: cardTypes[Math.floor(Math.random() * cardTypes.length)],
    };

    // 랜덤 족보 선택
    const pattern =
      patternNames[Math.floor(Math.random() * patternNames.length)];

    players.push({
      cards: [card1, card2],
      pattern: pattern,
    });
  }

  return {
    players: players,
    timestamp: new Date().toLocaleString(),
  };
}

File: Frontend/game/game.py
# game.py
import sys
import os
import random
import json
from datetime import datetime
from flask import Flask, request, jsonify, send_from_directory

# 경로 설정
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

app = Flask(__name__)

# 게임 기록 저장 (메모리 기반, 실제 서비스에서는 DB 사용 권장)
game_history = []

# 현재 게임 세션 정보
current_game_session = None

# 족보 점수 체계
pattern_scores = {
    # 특수 족보
    "38광땡": 200,
    "18광땡": 150,
    "13광땡": 130,
    
    "암행어사": 11,
    "땡잡이": 10,
    "구사": 90,
    "멍텅구리 구사":109,
    
    # 땡
    "장땡": 110,
    "9땡": 99,
    "8땡": 98,
    "7땡": 97,
    "6땡": 96,
    "5땡": 95,
    "4땡": 94,
    "3땡": 93,
    "2땡": 92,
    "1땡": 91,
    
    # 특수 조합
    "알리": 80,
    "독사": 70,
    "구삥": 60,
    "장삥": 50,
    "장사": 40,
    "세륙": 30,
    "갑오": 20,
    
    # 끗
    "8끗": 18,
    "7끗": 17,
    "6끗": 16,
    "5끗": 15,
    "4끗": 14,
    "3끗": 13,
    "2끗": 12,
    "1끗": 11,
    "망통": 10

}

# 카드 클래스
class Card:
    def __init__(self, month, type):
        self.month = month  # 1~10월
        self.type = type    # 1: 광,열끗 , 2: 띠
    
    def to_dict(self):
        return {
            'month': self.month,
            'type': self.type
        }

# 섯다 게임 클래스
class SutdaGame:
    def __init__(self, player_count):
        self.player_count = player_count
        self.deck = self._initialize_deck()
        self.players = []
        self._deal_cards()
    
    def _initialize_deck(self):
        # 1월부터 10월까지, 각 월마다 2장씩 카드 생성
        deck = []
        for month in range(1, 11):
            for type in range(1, 3):
                deck.append(Card(month, type))
        
        # 카드 섞기
        random.shuffle(deck) 
        return deck
    
    def _deal_cards(self):
        # 각 플레이어에게 2장씩 카드 배분
        for _ in range(self.player_count):
            if len(self.deck) < 2:
                # 카드가 부족하면 덱 초기화
                self.deck = self._initialize_deck()
            
            # 2장의 카드 뽑기
            cards = [self.deck.pop(), self.deck.pop()]
            pattern = self._calculate_pattern(cards)
            score = self._get_pattern_score(pattern)
            
            self.players.append({
                'cards': cards,
                'pattern': pattern,
                'score': score,
                'isWinner': False  # 승자 여부는 나중에 결정
            })

        # 승자 결정
        self._determine_winner()
    
    def _calculate_pattern(self, cards):
        # 카드 월 정보
        month1 = cards[0].month
        month2 = cards[1].month
        
        # 카드 타입 정보
        type1 = cards[0].type
        type2 = cards[1].type
        
        # 암행어사 (4월 열끗, 7월 열끗)
        if (month1 == 4 and month2 == 7 and type1 == 2 and type2 == 2) or (month1 == 7 and month2 == 4 and type1 == 2 and type2 == 2) :
            return "암행어사"
        
        # 땡잡이 (3월 광 , 7월 열끗)
        if (month1 == 3 and month2 == 7 and type1 == 1 and type2 == 2) or (month1 == 7 and month2 == 3 and type1 == 2 and type2 == 1) :
            return "땡잡이"
        
        # 38광땡 (3월, 8월 모두 광)
        if ((month1 == 3 and month2 == 8) or (month1 == 8 and month2 == 3)) and type1 == 1 and type2 == 1:
            return "38광땡"
        
        # 18광땡 (1월, 8월 모두 광)
        if ((month1 == 1 and month2 == 8) or (month1 == 8 and month2 == 1)) and type1 == 1 and type2 == 1:
            return "18광땡"
        
        # 13광땡 (1월, 3월 모두 광)
        if ((month1 == 1 and month2 == 3) or (month1 == 3 and month2 == 1)) and type1 == 1 and type2 == 1:
            return "13광땡"
        
        # 땡 (같은 월)
        if month1 == month2:
            if month1 == 10:
                return "장땡"
            else:
                return f"{month1}땡"
        
        # 알리 (1월, 2월)
        if (month1 == 1 and month2 == 2) or (month1 == 2 and month2 == 1):
            return "알리"
        
        # 독사 (1월, 4월)
        if (month1 == 1 and month2 == 4) or (month1 == 4 and month2 == 1):
            return "독사"
        
        # 구삥 (9월, 1월)
        if (month1 == 9 and month2 == 1) or (month1 == 1 and month2 == 9):
            return "구삥"
        
        # 장삥 (10월, 1월)
        if (month1 == 10 and month2 == 1) or (month1 == 1 and month2 == 10):
            return "장삥"
        
        # 장사 (10월, 4월)
        if (month1 == 10 and month2 == 4) or (month1 == 4 and month2 == 10):
            return "장사"
        
        # 세륙 (4월, 6월)
        if (month1 == 4 and month2 == 6) or (month1 == 6 and month2 == 4):
            return "세륙"
        
        # 구사,멍텅구리 구사 (4월, 9월),(4월 열끗, 9월 열끗)
        if (month1 == 4 and month2 == 9) or (month1 == 9 and month2 == 4):
            if (type1 == 2 and type2 == 2):
                return "멍텅구리 구사"
            return "구사"

        # 끗 (나머지)
        sum_value = (month1 + month2) % 10
        if sum_value == 9:
            return "갑오"
        elif sum_value == 0:
            return "망통"
        
        return f"{sum_value}끗"
    
    def to_dict(self):
        result = {
            'players': []
        }
        
        for player in self.players:
            player_dict = {
                'cards': [card.to_dict() for card in player['cards']],
                'pattern': player['pattern']
            }
            result['players'].append(player_dict)
        
        return result

    def _get_pattern_score(self, pattern):
        # 족보 점수 반환
        if pattern in pattern_scores:
            return pattern_scores[pattern]
        
        # 기본값 (알 수 없는 패턴)
        return 0
    
    def _determine_winner(self):
        # 각 플레이어의 점수 확인
        scores = [player['score'] for player in self.players]
        
        #특수 족보 적용용
        for player in self.players:
            if player['pattern'] == "암행어사":
                if any(op['pattern'] in ["18광땡", "13광땡"] for op in self.players):
                    player['score'] += 140
            elif player['pattern'] == "땡잡이":
                if any(op['pattern'] in ["9땡", "8땡", "7땡", "6땡", "5땡", "4땡", "3땡", "2땡", "1땡"] for op in self.players):
                    player['score'] += 90
        
        scores = [player['score'] for player in self.players]

        # 최고 점수 찾기
        max_score = max(scores) if scores else 0
        
        # 최고 점수를 가진 플레이어 수 확인 (무승부 체크)
        max_score_count = scores.count(max_score)
        
        # 무승부가 아닌 경우 승자 설정
        if max_score_count == 1:
            winner_index = scores.index(max_score)
            # ✅ 구사 또는 멍텅구리 구사일 경우 무승부 처리
            winner_pattern = self.players[winner_index]['pattern']
            if winner_pattern in ["구사", "멍텅구리 구사"]:
                return -2  # 구새패 재경기!

            self.players[winner_index]['isWinner'] = True
            return winner_index
        else:
            # 무승부
            return -1
    
    def to_dict(self):
        # 승자 결정
        winner_index = self._determine_winner()
        
        # 무승부 여부 확인
        if ( winner_index == -1):
            is_draw = 1 #무승부!
        elif ( winner_index == -2):
            is_draw = -1 #구사패 재경기!
        else:
            is_draw = 0
    
        result = {
            'players': [player for player in self.players],
            'winner': winner_index,
            'isDraw': is_draw,
            #"gusaRematch": gusa_rematch
        }
        
        # 플레이어 정보를 딕셔너리로 변환
        for i, player in enumerate(result['players']):
            result['players'][i] = {
                'cards': [card.to_dict() for card in player['cards']],
                'pattern': player['pattern'],
                'score': player['score'],
                'isWinner': player['isWinner']
            }
        
        return result

# 정적 파일 서빙
@app.route('/')
def index():
    return send_from_directory('static', 'index.html')

@app.route('/static/<path:path>')
def serve_static(path):
    return send_from_directory('static', path)

# API 엔드포인트: 게임 시작
@app.route('/api/game/start', methods=['POST'])
def start_game():
    global current_game_session
    
    data = request.json
    player_count = data.get('players', 1)
    
    # 플레이어 수 검증
    if player_count < 1 or player_count > 4:
        return jsonify({'error': 'Player count must be between 1 and 4'}), 400
    
    # 새 게임 생성
    game = SutdaGame(player_count)
    game_result = game.to_dict()
    
    # 타임스탬프 추가
    game_result['timestamp'] = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    
    # 현재 게임 세션 저장
    current_game_session = game_result
    
    # 게임 기록 저장
    save_game_history(game_result)
    
    return jsonify(game_result)

# API 엔드포인트: 게임 초기화 (카드 재배분 없음)
@app.route('/api/game/reset', methods=['POST'])
def reset_game():
    # 게임 세션 초기화만 수행 (카드 재배분 없음)
    return jsonify({'status': 'success', 'message': 'Game reset successfully'})

# API 엔드포인트: 게임 기록 조회
@app.route('/api/game/history', methods=['GET'])
def get_game_history():
    return jsonify(game_history)

# 게임 기록 저장 함수
def save_game_history(game_result):
    global game_history
    
    # 최대 10개까지만 저장
    if len(game_history) >= 10:
        game_history.pop(0)  # 가장 오래된 기록 제거
    
    game_history.append(game_result)

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=True)
